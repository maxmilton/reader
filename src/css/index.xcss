@import "./fonts.xcss";

// Packages
@import "@ekscss/framework";
@import "@ekscss/framework/addon/dark.xcss";

// Local
@import "./global.xcss";
