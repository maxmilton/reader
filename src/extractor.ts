/* eslint-disable unicorn/prefer-includes, unicorn/no-for-loop */

// Import html5parser from source directly to avoid tslib in their dist
import {
  parse,
  SyntaxKind,
  walk,
  type INode,
  type ITag,
} from 'html5parser/src';
import { create } from 'stage1';

interface Tag extends Omit<ITag, 'attributeMap'> {
  attributeMap: Record<string, string | undefined>;
}

const EXTRANEOUS_ELEMENTS = [
  '!--',
  'aside',
  'button',
  'canvas',
  'embed',
  'figcaption',
  'figure',
  'form',
  'head',
  'iframe',
  'input',
  'nav',
  'noscript',
  'script',
  'style',
  'svg',
  'textarea',
];
// FIXME: Needs more real-world testing as false positives are possible
//  ↳ Might need some kind of scoring logic to determine confidence
//  ↳ What if the user wants to read comments? Should the matching be different
//    if the input is the user selection?
const EXTRANEOUS_CLASSES = /comment|communit|contact|disqus|donat|extra|fundrais|meta|pager|pagination|popup|promo|related|remark|rss|share|shout|sidebar|sponsor|social|tags|tool|widget/i;
const BLOCK_ELEMENTS = [
  'address',
  'article',
  // 'aside',
  'blockquote',
  // 'canvas',
  'dd',
  'div',
  'dl',
  'fieldset',
  // 'figcaption',
  // 'figure',
  'footer',
  // 'form',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'header',
  'hgroup',
  'hr',
  'li',
  'main',
  // 'nav',
  // 'noscript',
  'ol',
  'output',
  'p',
  'pre',
  'section',
  'table',
  'tfoot',
  'ul',
];
const SKIP = true;

const textarea = create('textarea');

function decodeHTMLEntities(html: string) {
  textarea.innerHTML = html;
  return textarea.value;
}

function buildAttributeMap(node: ITag | Tag): asserts node is Tag {
  // eslint-disable-next-line no-param-reassign
  node.attributeMap = {};

  for (let index = 0; index < node.attributes.length; index++) {
    const attr = node.attributes[index];
    // eslint-disable-next-line no-param-reassign
    node.attributeMap[attr.name.value] = attr.value && attr.value.value;
  }
}

// Custom AST walker that can skip over subtrees
// https://github.com/lukeed/astray/blob/017484ce67402224304836e7d1a2fe2e116c3ae9/src/index.js
function walk2(
  node: INode[] | INode,
  parent: ITag,
  enter: (node: INode, parent: ITag) => void | typeof SKIP,
  leave: (node: INode) => void,
) {
  if (Array.isArray(node)) {
    for (let index = 0; index < node.length; index++) {
      walk2(node[index], parent, enter, leave);
    }
  } else {
    if (enter(node, parent) === SKIP) return;
    if (node.type === SyntaxKind.Tag && node.body) {
      walk2(node.body, node, enter, leave);
    }
    leave(node);
  }
}

/**
 * Attempt to extract the main content of a given HTML document.
 *
 * @param html - HTML markup. Since this is coming from the browser after the
 * page was rendered, it will be well-formed and safe.
 * @returns The main content of the document.
 */
export function extractText(html: string): string {
  const ast = parse(html);

  const idMap: Record<string, ITag> = {};
  const articles: ITag[] = [];
  const mains: ITag[] = [];
  let body: ITag;

  // First pass; populate attribute maps and collect references
  walk(ast, {
    enter: (node) => {
      if (node.type === SyntaxKind.Tag) {
        switch (node.name) {
          case 'article':
            articles.push(node);
            break;
          case 'body':
            body = node;
            break;
          case 'main':
            mains.push(node);
            break;
          default:
            break;
        }

        buildAttributeMap(node);

        const attrId = node.attributeMap.id;
        if (attrId) {
          idMap[attrId] = node;
        }
      }
    },
  });

  // Choose the best root node:
  //  1. <article> element if there is only one
  //  2. Element with id = article
  //  3. Element with id = post
  //  4. Element with id = content
  //  5. Element with id = main
  //  6. <main> element if there is only one
  //  7. Element with id = app
  //  8. Element with id = root
  //  9. <body> element (always defined)
  const root = articles.length === 1
    ? articles[0]
    : idMap.article
        || idMap.post
        || idMap.content
        || idMap.main
        || (mains.length === 1 ? mains[0] : idMap.app || idMap.root || body!);
  let text = '';

  // Second pass; clean up superfluous nodes and extract meaningful text
  walk2(
    root.body!,
    root,
    // eslint-disable-next-line consistent-return
    (node, parent) => {
      if (node.type === SyntaxKind.Tag) {
        if (
          EXTRANEOUS_ELEMENTS.indexOf(node.name) !== -1
          || (node.name === 'footer' && parent.name !== 'blockquote')
          // TODO: Fix types
          || ((node as unknown as Tag).attributeMap.class
            && EXTRANEOUS_CLASSES.test(
              (node as unknown as Tag).attributeMap.class!,
            ))
        ) {
          return SKIP;
        }
      } else {
        // Add text with consecutive whitespace collapsed
        text += (
          node.value.indexOf('&') !== -1
            ? decodeHTMLEntities(node.value)
            : node.value
        ).replace(/\s+/g, ' ');
      }
    },
    (node) => {
      if (
        node.type === SyntaxKind.Tag
        && BLOCK_ELEMENTS.indexOf(node.name) !== -1
      ) {
        // Add double space (which is turned into a newline later)
        text += '  ';
      }
    },
  );

  // console.log(stringify(root, html));
  // console.log(root);

  return (
    text
      .trim()
      // ensure single consecutive \n padded with space
      .replace(/[\n ]{2,}/g, ' \n ')
      // fix missing space around em dashes
      .replace(/(\S)—(\S)/g, '$1 — $2')
  );
}

// // Simple stringify AST to prettified HTML-like structure for debugging
// function stringify(node: INode, html: string, level = 1): string {
//   if (node.type === SyntaxKind.Text) return node.value.replace(/\s+/g, ' ');
//   if (node.name === '!--') return html.slice(node.start, node.end);

//   const attrs = node.attributes
//     .map((attr) => html.slice(attr.start, attr.end))
//     .join(' ');
//   const head = `<${node.rawName}${attrs ? ` ${attrs}` : ''}>`;

//   if (!node.body || node.body.length === 0) return head;

//   /* eslint-disable prefer-template */ // template string breaks after minification
//   return (
//     head
//     + '\n'
//     + '  '.repeat(level)
//     + node.body
//       .filter((n) => !(n.type === SyntaxKind.Text && n.value.trim() === ''))
//       .map((n) => stringify(n, html, level + 1))
//       .join('\n' + '  '.repeat(level))
//     + '\n'
//     + '  '.repeat(level - 1)
//     + '</'
//     + node.rawName
//     + '>'
//   );
// }
