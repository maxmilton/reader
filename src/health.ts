/**
 * Tracks JavaScript exceptions using the BugBox client.
 *
 * @fileoverview Chrome Extensions v3 disallow remote code execution,
 * so loading BugBox from a CDN isn't an option. Using a local version
 * is both faster and compliant. Since we're targeting modern browsers,
 * the lite client is sufficient.
 */

// TODO: Use bugbox/lite client once available.
import * as bugbox from 'bugbox';

chrome.tabs.query({ active: true, currentWindow: true }, ([tab]) => {
  bugbox.meta.url = tab.url;
});
