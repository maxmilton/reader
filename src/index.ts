// FIXME: Must come first; prevent import reordering. DO NOT ORGANIZE IMPORTS.
//  â†³ https://github.com/microsoft/TypeScript/issues/41494
import { append, createFragment, setupSyntheticEvent } from 'stage1';
import { Footer } from './components/Footer';
import { Reader } from './components/Reader';
import './css/index.xcss';

declare global {
  interface HTMLElement {
    /** `stage1` synthetic click event handler. */
    __click?(event: MouseEvent): void;
  }
}

const frag = createFragment();

append(Reader(), frag);
append(Footer(), frag);
append(frag, document.body);

setupSyntheticEvent('click');
