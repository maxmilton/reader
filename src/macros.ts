// FIXME: Remove this file once no longer needed.

// FIXME: This is a convoluted workaround for a bug in the bun macro system,
// where it crashes when doing string literal template interpolation. See:
// https://github.com/oven-sh/bun/issues/3641
export function interpolate(text: string, values: string[]): string {
  let result = text;

  // eslint-disable-next-line unicorn/no-array-for-each
  values.forEach((value, index) => {
    result = result.replace(new RegExp(`%%${String(index + 1)}%%`, 'g'), value);
  });

  return result;
}
