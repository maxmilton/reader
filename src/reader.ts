import './css/index.xcss';

import { append, createFragment, setupSyntheticEvent } from 'stage1';
import { Footer } from './components/Footer';
import { Reader } from './components/Reader';

declare global {
  interface HTMLElement {
    /** `stage1` synthetic click event handler. */
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
    __click?(event: MouseEvent): void | false | Promise<void>;
  }
}

const frag = createFragment();

append(Reader(), frag);
append(Footer(), frag);
append(frag, document.body);

setupSyntheticEvent('click');
