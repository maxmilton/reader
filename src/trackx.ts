/**
 * Track JS exceptions with the trackx client.
 *
 * @fileoverview In Chrome extensions v3 running remote code is not allowed so
 * trackx via CDN would not work + loading local code is obviously much faster.
 *
 * Since the runtime environment will be modern browsers the lite client is OK.
 */

import * as trackx from 'trackx/lite';

trackx.setup('https://api.trackx.app/v1/9lbe1l9le4x');
trackx.meta.release = process.env.APP_RELEASE;

if (process.env.NODE_ENV !== 'production') {
  trackx.meta.NODE_ENV = process.env.NODE_ENV;
}

void chrome.tabs.query({ active: true, currentWindow: true }).then(([tab]) => {
  trackx.meta.url = tab.url;
});

void fetch('https://api.trackx.app/v1/9lbe1l9le4x/ping', {
  method: 'POST',
  keepalive: true,
  mode: 'no-cors',
});
