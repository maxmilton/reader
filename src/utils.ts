export async function getCurrentTab(): Promise<chrome.tabs.Tab> {
  const [tab] = await chrome.tabs.query({ active: true, currentWindow: true });
  return tab;
}

/**
 * Execute a function in the context of the current active tab.
 *
 * Note: The function will be serialised + injected into the tab's page and
 * then run as a content script so it must be self contained (a pure function;
 * no access or mutation outside its own closure scope).
 *
 * @param func - Function to run in page. Must be self-contained.
 * @param args - Arguments to pass to function. Must be JSON-serializable.
 *
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#method-executeScript
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#type-ScriptInjection
 * @see https://developer.chrome.com/en/blog/crx-scripting-api/#injecting-a-function-with-arguments
 */
export async function exec<T, A extends any[] = []>(
  func: (..._args: A) => T,
  args?: A,
): Promise<T> {
  const tab = await getCurrentTab();

  const [{ result }] = await chrome.scripting.executeScript({
    target: { tabId: tab.id! },
    func,
    args,
  });

  return result as T;
}

/**
 * Get the Optimal Recognition Point (OPR) index for a given word phrase.
 */
export function indexOfORP(word: string): number {
  let len = word.length;

  while (/\W/.test(word[--len]));

  if (/\W/.test(word[0])) len++;

  switch (++len) {
    case 0:
    case 1: {
      return 0;
    }
    case 2:
    case 3: {
      return 1;
    }
    default: {
      return Math.floor(len / 2) - 1;
    }
  }
}
