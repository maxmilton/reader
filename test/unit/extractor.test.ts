import { describe, expect, test } from 'bun:test';
import { extractText } from '../../src/extractor';

const minimal = '<html><body>x</body></html>';
const basic = await Bun.file('test/unit/fixtures/basic.html').text();
const broken = await Bun.file('test/unit/fixtures/broken.html').text();
const wikipediaSimple = await Bun.file('test/unit/fixtures/wikipedia-simple.html').text();
const wikipedia = await Bun.file('test/unit/fixtures/wikipedia.html').text();

// TODO: Test each of the root element selection logic branches

test('is a function', () => {
  expect(extractText).toBeFunction();
});

test('takes a single argument', () => {
  expect(extractText).toHaveLength(1);
});

test('returns a string', () => {
  const result = extractText(minimal);
  expect(result).toBeString();
});

describe('basic.html', () => {
  test('returns expected text', () => {
    const result = extractText(basic);
    expect(result).toBe('Basic HTML \n Some text.');
  });
});

describe('broken.html', () => {
  test('throws when processing', () => {
    expect(() => extractText(broken)).toThrow();
  });
});

describe('wikipedia-simple.html', () => {
  // FIXME: Don't skip this test. Test snapshots are currently broken in bun.
  test.skip('returns expected text', () => {
    const result = extractText(wikipediaSimple);
    expect(result).toMatchSnapshot();
  });
});

describe('wikipedia.html', () => {
  // FIXME: Don't skip this test. Test snapshots are currently broken in bun.
  test.skip('returns expected text', () => {
    const result = extractText(wikipedia);
    expect(result).toMatchSnapshot();
  });
});
