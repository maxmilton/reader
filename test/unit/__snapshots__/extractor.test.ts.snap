// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`wikipedia-simple.html returns expected text 1`] = `
"\n\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\tToggle the table of contents\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\tTest 
 \n\t\t\t\t\t\t\t\n\n\t\n\t\n\n6 languages\n\t\n\t\n\n\t\t\n\t\t\t\n\t\t\t\n\t\t\t\t\n\t\t\t\tDeutsch 
 Italiano 
 עברית 
 Русский 
 Српски / srpski 
 Українська 
 \n\t\t\t 
 \n\t\t\tChange links 
 \n\t\t 
 \n\n\t 
 \n 
 \n 
 \n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t 
 \n\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\t\n\t\t 
 \n\n\t\t\t\t\t\tFrom Simple English Wikipedia, the free encyclopedia 
 \n\t\t\t\t\t 
 \n\t\t\t\t\t 
 \n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\nA test is a way of checking something to see if it is true, or false, or if it is edible or not. If something can be tested, or finishes the tests correctly, it is testable. The Concise Oxford English Dictionary defines a test as: "a procedure intended to establish the quality, performance, or reliability of something".[1]\n 
 A test is different from an experiment: Before a test is done, there is an expected result. The test is performed, to show this result. In an experiment, the outcome is open. Very often, tests are performed as part of an experiment.\n 
 \n\nProducts[change | change source] 
 \nProducts are usually tested for quality, so customers will get good products.\n 
 In software engineering, a test is used to see if the software system can do what it should. Software is tested before it is released. Alpha testing is where software developers check the software for bugs. Software can also be checked for quality and usability. Beta testing is done by groups of users.\n 
 Tests of cars and other vehicles include a crash test. The car is put under severe conditions to see what will make it fail, or deliberately crashed to measure the damage. Other machines can also be crash tested. Crash test dummies can be used instead of humans. They are placed in the car seat to see if a human in the crash would have been injured or killed.\n 
 \nPeople[change | change source] 
 \n\nPeople are tested to see what they have learned. This is often called an assessment or examination. In learning, a test item is a question, or set of questions.\n 
 Many people think tests are valuable. They believe tests:\n 
 \nare a quick and fair way of judging a test taker's performance 
 \nenable predictions about test takers to be made 
 \nallow selection 
 \nimprove performance by highlighting areas that need work. 
 \nHowever, academic tests are not perfect measures. Tests could only partly measure a student’s memory and maybe their understanding. The test would only be about a small part of the subject, and only at that moment in time. \n 
 People, animals and plants can also be tested for illnesses. For example, a blood test can be used to check for disease.\n 
 \nScience[change | change source] 
 \nIn science, tests can done to check for a presence of a substance, or to check the quality of something.\n 
 \nReferences[change | change source] 
 \n\n↑ Stevenson, Angus; Waite, Maurice (2011). Concise Oxford English Dictionary (Twelfth ed.). Oxford University Press. p. 1490. ISBN 978-0-19-960108-0.\n 
 \n 
 \nOther websites[change | change source] 
 \nTests and Experiments: Similarities and Differences Archived 2016-03-04 at the Wayback Machine 
 \n\x3C!-- \nNewPP limit report\nParsed by mw2390\nCached time: 20230718002129\nCache expiry: 1814400\nReduced expiry: false\nComplications: [vary‐revision‐sha1, show‐toc]\nCPU time usage: 0.078 seconds\nReal time usage: 0.112 seconds\nPreprocessor visited node count: 91/1000000\nPost‐expand include size: 1841/2097152 bytes\nTemplate argument size: 0/2097152 bytes\nHighest expansion depth: 5/100\nExpensive parser function count: 0/500\nUnstrip recursion depth: 1/20\nUnstrip post‐expand size: 2709/5000000 bytes\nLua time usage: 0.047/10.000 seconds\nLua memory usage: 2515437/52428800 bytes\nNumber of Wikibase entities loaded: 0/400\n-->\n\x3C!--\nTransclusion expansion time report (%,ms,calls,template)\n100.00% 86.090 1 -total\n 82.95% 71.414 1 Template:Cite_book\n 15.53% 13.371 1 Template:Webarchive\n-->\n\n\x3C!-- Saved in parser cache with key simplewiki:pcache:idhash:823-0!canonical and timestamp 20230718002129 and revision id 8058954. Rendering was triggered because: page-view\n -->\n 
 \x3C!--esi 
 -->\nRetrieved from "https://simple.wikipedia.org/w/index.php?title=Test&oldid=8058954" 
 \n\t\t\t\t\tCategories: Basic English 850 words 
 Tests 
 Hidden category: Webarchive template wayback links 
 \n\t\t\t\t 
 \n\t\t\t"
`;
